lib/std/crypto/bcrypt.zig

```diff
@@ -213,84 +275,205 @@ fn strHashInternal(password: []const u8, rounds_log: u6, salt: [salt_length]u8)
+        }
+        const params = try parseParams(str[0..7]);
+        const salt_str = str[7..][0..salt_str_length];
+        const derived_key_str = str[7 + salt_str_length ..][0..derived_key_str_length];
+
+        var salt: [salt_length]u8 = undefined;
+        try Codec.decode(&salt, salt_str);
+
+        var dk: [derived_key_length]u8 = undefined;
+        kdf(&dk, password, salt, params);
+
+        var derived_key: [derived_key_str_length]u8 = undefined;
+        Codec.encode(&derived_key, dk[0 .. dk.len - 1]);
+
+        const passed = crypto.utils.timingSafeEql(
+            [derived_key_str_length]u8,
+            derived_key,
+            derived_key_str.*,
+        );
+        crypto.utils.secureZero(u8, &derived_key);
+        if (!passed) {
+            return Error.PasswordVerificationFailed;
+        }
+    }
+
+    /// Derive key from password and return crypt encoded string
+    pub fn create(password: []const u8, params: Params) [pwhash_str_length]u8 {
+        var salt_bin: [salt_length]u8 = undefined;
+        crypto.random.bytes(&salt_bin);
+
+        var dk: [derived_key_length]u8 = undefined;
+        kdf(&dk, password, salt_bin, params);
+
+        var salt: [salt_str_length]u8 = undefined;
+        Codec.encode(&salt, &salt_bin);
+
+        var derived_key: [derived_key_str_length]u8 = undefined;
+        Codec.encode(&derived_key, dk[0 .. dk.len - 1]);
+
+        var buf: [pwhash_str_length]u8 = undefined;
+        const s = fmt.bufPrint(
+            &buf,
+            "{s}b${d}{d}${s}{s}",
+            .{
+                prefix,
+                params.log_rounds / 10,
+                params.log_rounds % 10,
+                salt,
+                derived_key,
+            },
+        ) catch unreachable;
+        debug.assert(s.len == buf.len);
+        return buf;
+    }
+};
+
+pub const HashOptions = struct {
+    kdf_params: Params,
+    encoding: crypto.pwhash.Encoding,
+};
 
-/// Compute a hash of a password using 2^rounds_log rounds of the bcrypt key stretching function.
+/// Compute a hash of a password using 2^log_rounds rounds of the bcrypt key stretching function.
 /// bcrypt is a computationally expensive and cache-hard function, explicitly designed to slow down exhaustive searches.
 ///
 /// The function returns a string that includes all the parameters required for verification.
 ///
 /// IMPORTANT: by design, bcrypt silently truncates passwords to 72 bytes.
 /// If this is an issue for your application, hash the password first using a function such as SHA-512,
 /// and then use the resulting hash as the password parameter for bcrypt.
-pub fn strHash(password: []const u8, rounds_log: u6) Error![hash_length]u8 {
-    var salt: [salt_length]u8 = undefined;
-    crypto.random.bytes(&salt);
-    return strHashInternal(password, rounds_log, salt);
+pub fn strHash(
+    allocator: *mem.Allocator,
+    password: []const u8,
+    options: HashOptions,
+    out: []u8,
+) (Error || mem.Allocator.Error)![]u8 {
```

Comment:
You are using `Error || mem.Allocator.Error` 9 times.
Could you extract/name it?